spring-ai-core/src/test/java/org/springframework/ai/chat/ChatBuilderTests.java

```diff
@@ -0,0 +1,134 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.junit.jupiter.api.Test;
+
+import org.springframework.ai.chat.prompt.ChatOptions;
+import org.springframework.ai.chat.prompt.ChatOptionsBuilder;
+import org.springframework.ai.chat.prompt.Prompt;
+import org.springframework.ai.model.function.FunctionCallback;
+import org.springframework.ai.model.function.FunctionCallbackWrapper;
+import org.springframework.ai.model.function.FunctionCallingOptions;
+import org.springframework.ai.model.function.FunctionCallingOptionsBuilder;
+
+/**
+ * Unit Tests for {@link Prompt}.
+ *
+ * @author youngmon
+ * @since 0.8.1
+ */
+public class ChatBuilderTests {
+
+	@Test
+	void createNewChatOptionsTest() {
+		Float temperature = 1.1f;
+		Float topP = 2.2f;
+		Integer topK = 111;
+
+		ChatOptions options = ChatOptionsBuilder.builder()
+			.withTemperature(temperature)
+			.withTopK(topK)
+			.withTopP(topP)
+			.build();
+
+		assertThat(options.getTemperature()).isEqualTo(temperature);
+		assertThat(options.getTopP()).isEqualTo(topP);
+		assertThat(options.getTopK()).isEqualTo(topK);
+	}
+
+	@Test
+	void duplicateChatOptionsTest() {
+		Float initTemperature = 1.1f;
+		Float initTopP = 2.2f;
+		Integer initTopK = 111;
+
+		ChatOptions options = ChatOptionsBuilder.builder()
+			.withTemperature(initTemperature)
+			.withTopP(initTopP)
+			.withTopK(initTopK)
+			.build();
+
+		Integer newTopK = 222;
+
+		ChatOptions newOptions = ChatOptionsBuilder.builder(options)
+			// setTopK
+			.withTopK(newTopK)
+			.build();
+
+		assertThat(newOptions.getTemperature()).isEqualTo(initTemperature);
+		assertThat(newOptions.getTopP()).isEqualTo(initTopP);
+		assertThat(newOptions.getTopK()).isEqualTo(newTopK);
+	}
+
+	@Test
+	void createFunctionCallingOptionTest() {
+		Float temperature = 1.1f;
+		Float topP = 2.2f;
+		Integer topK = 111;
+		List<FunctionCallback> functionCallbacks = new ArrayList<>();
+		Set<String> functions = new HashSet<>();
+
+		String func = "func";
+		FunctionCallback cb = FunctionCallbackWrapper.<Integer, Integer>builder(i -> i)
+			.withName("cb")
+			.withDescription("cb")
+			.build();
+
+		functions.add(func);
+		functionCallbacks.add(cb);
+
+		FunctionCallingOptions options = FunctionCallingOptionsBuilder.builder()
+			.withFunctionCallbacks(functionCallbacks)
+			.withFunctions(functions)
+			.withTopK(topK)
+			.withTopP(topP)
+			.withTemperature(temperature)
+			.build();
+
+		// Callback Functions
+		assertThat(options.getFunctionCallbacks()).isNotNull();
+		assertThat(options.getFunctionCallbacks().size()).isEqualTo(1);
+		assertThat(options.getFunctionCallbacks().contains(cb));
+
+		// Functions
+		assertThat(options.getFunctions()).isNotNull();
+		assertThat(options.getFunctions().size()).isEqualTo(1);
+		assertThat(options.getFunctions().contains(func));
+
+		// Immutable
+		options.getFunctionCallbacks().add(cb);
```

Comment:
If we want the created options class to be immutable then we should create a non modifiable collection and throw an exception.  We would also need to ensure that all the other options classes behave the same way.